woocommerce/woocommerce-android
# Fix tracking order details events multiple times

Closes: #4779 


### Description

This PR makes `OrderDetailRepository` to be injected in `ViewModelScope` and moves analytics tracking from `Repository` to `ViewModel` for Order Details screen. 

It also unregisters `OrderDetailRepository` from EventBus dispatcher, allowing garbage collector to collect those objects.

This will assert that:
1. for every object that will use `OrderDetailRepository` in `OrderDetailViewModel` only one instance of repository will be created and released in `OrderDetailViewModel` too (case of `CardReaderPaymentCollectibilityChecker`)
2. For nested screens (like `CreateShippingLabel`) a new instance of repository will be created (so `continuation`s for different ViewModels) will work independently but as we moved sending analytics events to `OrderDetailViewModel` which is always one for details screen - the events will be send only once.

For more detailed explanations, please visit the Slack thread: p1631609358257000/1631557182.228900-slack-C6H8C3G23

This is how it **did** look for "change order status" action.
![image](https://user-images.githubusercontent.com/5845095/133259196-f25294d7-48c4-423c-b4eb-c5731a4795b1.png)

(these logs were added locally, don't expect them to see on your build).


### Testing instructions

Before testing this change you might want to reproduce the issue on your own. Please see #4779 for instructions how to do it.

#### Test sending multiple events fix

1. Open the app
2. Go to list of orders
3. Go to details of any order
4. Go back to list
5. Repeat 3 and 4 for a few times (like 5 times)
6. Go again to detail of any order
7. Change the status of the order
8. **Assert there's exactly one `order_status_change_success` event logged**

#### Test fix for leak of `OrderDetailRepository`

1. Apply PATCH from the bottom of PR description. It adds logs for better visibility
2. Run the app
3. Open logs. You can filter by `PR-4780`
4. Open details of any order. **Assert there's log "Creating instance"**
5. Close details of any order. **Assert there's log "Destructing instance" with the same object reference as above**
6. Perform 4 and 5 several times.

Perform above steps for other screens like `Issue refund` or `Create shipping label`.

---

```
Index: WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailRepository.kt
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
diff --git a/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailRepository.kt b/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailRepository.kt
--- a/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailRepository.kt	(revision 93ab9093138feb43ab792a024edb9887c93e1662)
+++ b/WooCommerce/src/main/kotlin/com/woocommerce/android/ui/orders/details/OrderDetailRepository.kt	(date 1631710984335)
@@ -35,6 +35,7 @@
 import org.wordpress.android.fluxc.store.*
 import org.wordpress.android.fluxc.store.WCOrderStore.*
 import org.wordpress.android.fluxc.store.WCProductStore.OnProductChanged
+import org.wordpress.android.util.AppLog
 import javax.inject.Inject
 
 @ViewModelScoped
@@ -56,9 +57,14 @@
     private var continuationDeleteShipmentTracking = ContinuationWrapper<Boolean>(ORDERS)
 
     init {
+        AppLog.d(AppLog.T.API, "PR-4780 Creating instance: $this")
         dispatcher.register(this)
     }
 
+    protected fun finalize() {
+        AppLog.d(AppLog.T.API, "PR-4780 Destructing instance: $this")
+    }
+
     fun onCleanup() {
         dispatcher.unregister(this)
     }

```
